flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/AbstractUserClassPathJobGraphRetriever.java

```diff
@@ -0,0 +1,115 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.entrypoint.component;
+
+import org.apache.flink.core.fs.Path;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.FileVisitOption;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ *  Abstract class for the JobGraphRetriever, which wants to get classpath user's code depends on.
+ */
+
+public abstract class AbstractUserClassPathJobGraphRetriever implements JobGraphRetriever {
+
+	protected static final Logger LOG = LoggerFactory.getLogger(AbstractUserClassPathJobGraphRetriever.class);
+
+	public static final String DEFAULT_JOB_DIR = "job";
+
+	/** The directory contains all the jars, which user code depends on. */
+	@Nullable
+	private final String jobDir;
+
+	private List<URL> userClassPaths;
+
+	public AbstractUserClassPathJobGraphRetriever(String jobDir) {
+		this.jobDir = jobDir;
+	}
+
+	public List<URL> getUserClassPaths() throws IOException {
+		if (userClassPaths == null) {
+			userClassPaths = scanJarsInJobClassDir(jobDir);
+		}
+		return userClassPaths;
+	}
+
+	private List<URL> scanJarsInJobClassDir(String dir) throws IOException {
+
+		if (dir == null) {
+			return Collections.emptyList();
+		}
+
+		final File dirFile = new File(new Path(dir).toString());
+		final List<URL> jarURLs = new LinkedList<>();
+
+		if (!dirFile.exists()) {
+			LOG.warn("the job dir " + dirFile + " dose not exists.");
+			return Collections.emptyList();
+		}
+		if (!dirFile.isDirectory()) {
+			LOG.warn("the job dir " + dirFile + " is not a directory.");
+			return Collections.emptyList();
+		}
+
+		Files.walkFileTree(dirFile.toPath(),
+			EnumSet.of(FileVisitOption.FOLLOW_LINKS),
+			Integer.MAX_VALUE,
+			new SimpleFileVisitor<java.nio.file.Path>(){
+
+			@Override
+			public FileVisitResult visitFile(java.nio.file.Path file, BasicFileAttributes attrs)
+				throws IOException {
+				FileVisitResult fileVisitResult = super.visitFile(file, attrs);
+				if (file.getFileName().toString().endsWith(".jar")) {
+					LOG.info("add " + file.toString() + " to user classpath");
+					if (file.isAbsolute()) {
+						jarURLs.add(file.toUri().toURL());
+					} else {
+						jarURLs.add(
+							new URL(new URL(file.getFileName().toUri().getScheme() + ":"), file.toString())
```

Comment:
Do you have a formal reference for that why a `<schema>:<relativePath>` could work?
Could you have comments for what hope to generate here?

Reply:
https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#URL(java.net.URL,%20java.lang.String)

> If the spec's path component begins with a slash character "/" then the path is treated as absolute and the spec path replaces the context path.
Otherwise, the path is treated as a relative path and is appended to the context path, as described in RFC2396.

This constructor uses the `context` and `spec` to build a URL.  `context` is used as a default URL information if "spec" does not specify.

Reply:
> MalformedURLException - if no protocol is specified, or an unknown protocol is found, or spec is null.

Ok. Looks the main reason is the protocol needs to be specified to avoid `MalformedURLException`.

Reply:
Can't we say that this method only reads local directories and list local files? Then we would not have to deal with URLs and could simply return a `File` here.

Reply:
If we need to support remote files, then I would suggest to use `Path` instead of URLs.